functorch/csrc/BatchRulesRandomness.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <ATen/ATen.h>
+#include <functorch/csrc/DynamicLayer.h>
+#include <functorch/csrc/BatchRulesHelper.h>
+
+namespace at {
+namespace functorch {
+
+void check_randomness(std::string randomness) {
+  TORCH_CHECK(
+    randomness != "error",
+    "vmap: called random operation while in randomness error mode. Please either use the "
+    "'same' or 'different' randomness flags on vmap or perform the randomness operation out of vmap"
+  );
+}
+
+template <typename F, F Func, typename... ExtraArgs>
+Tensor random_batching_rule(IntArrayRef shape, ExtraArgs... extra_args) {
+  c10::impl::ExcludeDispatchKeyGuard guard(kVmapModeKey);
+  auto maybe_layer = maybeCurrentDynamicLayer();
+  VmapDimVector shapeVec(shape.begin(), shape.end());
+  shapeVec.insert(shapeVec.begin(), maybe_layer->batchSize());
+  std::string randomness = maybe_layer->randomness();
+  check_randomness(randomness);
+  if (randomness == "different") {
+    return makeBatched(Func(shapeVec, std::forward<ExtraArgs>(extra_args)...), 0, maybe_layer->layerId());
+  } else {
+    const auto res = Func(shape, std::forward<ExtraArgs>(extra_args)...);
+    return makeBatched(res.unsqueeze(0).expand(shapeVec), 0, maybe_layer->layerId());
```

Comment:
@samdow to improve performance this should return a regular Tensor, not a BatchedTensor with an expanded underlying Tensor.

Let's say we did an operation on the resulting Tensor, like x.sin(). If the tensor was expanded, we'll end up creating a new contiguous tensor and materializing all of the memory. If the Tensor wasn't expanded, then we save on some performance

Reply:
Ah I see, it was fixed in https://github.com/pytorch/functorch/pull/410/files. You read my mind, three weeks ago :)

Reply:
Yeah this was actually a rebase error--I meant to change it in this one too 🤦‍♀️